sopa/embedding/patches.py

```diff
@@ -0,0 +1,133 @@
+import logging
+from typing import Callable
+
+import dask as da
+import numpy as np
+import tqdm
+from spatial_image import SpatialImage
+from spatialdata import SpatialData, bounding_box_query
+from spatialdata.models import Image2DModel
+from spatialdata.transformations import Scale
+
+from sopa._sdata import get_key
+from sopa.segmentation import Patches2D
+
+log = logging.getLogger(__name__)
+
+
+def _get_best_level_for_downsample(level_downsamples, downsample, epsilon=0):
+    """return the best level for a given downsampling factor"""
+    if downsample <= 1.0:
+        return 0
+    for level, ds in enumerate(level_downsamples):
+        if ds > downsample + epsilon:
+            return level - 1
+    return len(level_downsamples) - 1
+
+
+def _get_extraction_parameters(tiff_metadata, target_magnification, patch_width):
+    if tiff_metadata["properties"].get("tiffslide.objective-power"):
+        downsample = (
+            int(tiff_metadata["properties"].get("tiffslide.objective-power")) / target_magnification
+        )
+    elif tiff_metadata["properties"].get("tiffslide.mpp-x"):
+        obj2mpp = {80: 0.125, 40: 0.25, 20: 0.5, 10: 1.0, 5: 2.0}
+        mppdiff = []
+        for objpow, mpp in obj2mpp.items():
+            mppdiff += [abs(mpp - float(tiff_metadata["properties"].get("tiffslide.mpp-x")))]
+        idx = np.argmin([abs(mpp - 0.44177416504682804) for objpow, mpp in obj2mpp.items()])
+        mpp_obj = list(obj2mpp.keys())[idx]
+        downsample = mpp_obj / target_magnification
+    else:
+        log.error("Error retrieving the mpp for {{image_key}}, skipping tile extration.")
```

Comment:
Here `image_key` is not accessible, so we can't log it

Reply:
What if I put this log after the call of the function and check if it returns false, instead of putting it here?

Reply:
Yes, sounds good! Indeed it will avoid passing the argument to this function